src/Microsoft.Diagnostics.Monitoring.RestServer/DiagnosticServices.cs

```diff
@@ -141,41 +147,47 @@ public async Task<IProcessInfo> GetProcessAsync(ProcessKey? processKey, Cancella
                 }
             }
 
-            return await GetSingleProcessInfoAsync(
-                endpointInfos,
+            return GetSingleProcessInfo(
+                processInfos,
                 processKey: null);
         }
 
-        private async Task<IProcessInfo> GetSingleProcessInfoAsync(IEnumerable<IEndpointInfo> endpointInfos, ProcessKey? processKey)
+        private IProcessInfo GetSingleProcessInfo(IEnumerable<IProcessInfo> processInfos, ProcessKey? processKey)
         {
             if (processKey.HasValue)
             {
                 if (processKey.Value.RuntimeInstanceCookie.HasValue)
                 {
                     Guid cookie = processKey.Value.RuntimeInstanceCookie.Value;
-                    endpointInfos = endpointInfos.Where(info => info.RuntimeInstanceCookie == cookie);
+                    processInfos = processInfos.Where(info => info.EndpointInfo.RuntimeInstanceCookie == cookie);
                 }
 
                 if (processKey.Value.ProcessId.HasValue)
                 {
                     int pid = processKey.Value.ProcessId.Value;
-                    endpointInfos = endpointInfos.Where(info => info.ProcessId == pid);
+                    processInfos = processInfos.Where(info => info.EndpointInfo.ProcessId == pid);
+                }
+
+                string processName = processKey.Value.ProcessName;
+                if (!string.IsNullOrEmpty(processName))
+                {
+                    processInfos = processInfos.Where(info => string.Equals(info.ProcessName, processName, StringComparison.OrdinalIgnoreCase));
                 }
             }
 
-            IEndpointInfo[] endpointInfoArray = endpointInfos.ToArray();
-            switch (endpointInfoArray.Length)
+            IProcessInfo[] processInfosArray = processInfos.ToArray();
+            switch (processInfosArray.Length)
             {
                 case 0:
                     throw new ArgumentException("Unable to discover a target process.");
                 case 1:
-                    return await ProcessInfo.FromEndpointInfoAsync(endpointInfoArray[0]);
+                    return processInfosArray[0];
                 default:
 #if DEBUG
-                    IEndpointInfo endpointInfo = endpointInfoArray.FirstOrDefault(info => string.Equals(Process.GetProcessById(info.ProcessId).ProcessName, "iisexpress", StringComparison.OrdinalIgnoreCase));
-                    if (endpointInfo != null)
+                    IProcessInfo processInfo = processInfosArray.FirstOrDefault(info => string.Equals(Process.GetProcessById(info.EndpointInfo.ProcessId).ProcessName, "iisexpress", StringComparison.OrdinalIgnoreCase));
```

Comment:
Were you thinking that we would further expand ProcessKey to support other scenarios as well? For example, I might want to specify the metrics default by using command line substring, since the most common process name of dotnet.exe won't work very well.

Reply:
Sure, we can add a "command line contains the following list of values" filter in the future. I would say that is something that would NOT be available in the type conversion for use in the path/query though.